### ISSUE TITLE
EnumSet<> options

### ISSUE THREAD

**Issue by leeatkinson** [remkop/picocli#628]
I have a enum and I want an argument to be an EnumSet so I can set flags on the CLI, for example:
```
foo --result-types PARTIAL,COMPLETE
```
with code similar to:
```
public enum ResultTypes {
    NONE,
    PARTIAL,
    COMPLETE
}

@Option(names = "--result-types")
private EnumSet<ResultTypes> resultTypes = EnumSet.of(ResultTypes.COMPLETE);
```
But I get the error when executed:
```
Invalid value for option '--result-types' (<ResultTypes>): expected one of [NONE, PARTIAL, COMPLETE] (case-sensitive) but was 'PARTIAL,COMPLETE'
```
I have written a converter to parse the string:
```
public class ResultTypesEnumSetConverter implements CommandLine.ITypeConverter<EnumSet<ResultTypes>> {
    @Override
    public EnumSet<ResultTypes> convert(String value) throws Exception {
        String[] split = value.split(",");
        EnumSet<ResultTypes> types = EnumSet.noneOf(ResultTypes.class);
        for (int i = 0; i < split.length; i++) {
            types.add(ResultTypes.valueOf(split[i]));
        }
        return types;
    }
}
```
but this doesn't seem to work:
```
InstantiationException: java.util.EnumSet while processing argument at or before arg[1] 'PARTIAL,COMPLETE'
```

**Comment by remkop** [comment]
Can you run again with `-Dpicocli.trace=DEBUG` and show the output?

**Comment by remkop** [comment]
Thank you for raising this issue! I was able to reproduce the problem, please ignore my previous comment.

This is a shortcoming in the current version of picocli.

Both problems have the same cause: picocli detects that `EnumSet<ResultTypes>` is a `Collection` that can hold multiple `ResultTypes` instances. If the original field value is `null` or has a non-empty collection as it initial/default value, it tries to create an empty instance of the collection and add the `ResultTypes` instances found during parsing to this new collection instance. The idea is that when done, it will replace the field value with the new collection. 

The problem is that picocli currently has no special treatment for `EnumSet`, and there is no default constructor so the instantiation fails. Even with the custom type converter, picocli will try to instantiate the empty EnumSet...

At the moment the workaround is to either:
* use a different `Set` implementation like HashSet, TreeSet or LinkedHashSet
* initialize the field with an _empty_ EnumSet with `EnumSet.noneOf(ResultTypes.class)`

With that last workaround, this test passes:
```java
enum ResultTypes {
    NONE,
    PARTIAL,
    COMPLETE
}
@Test
public void testEnumSet() {
    class App {
        @Option(names = "--result-types", split = ",") //, converter = ResultTypesEnumSetConverter.class)
        //private EnumSet<ResultTypes> resultTypes = EnumSet.of(ResultTypes.COMPLETE);
        private EnumSet<ResultTypes> resultTypes = EnumSet.noneOf(ResultTypes.class);
    }
    App app = new App();
    new CommandLine(app).parseArgs("--result-types", "PARTIAL,COMPLETE");

    assertEquals(EnumSet.of(PARTIAL, COMPLETE), app.resultTypes);
}
```
I am a big fan of using `EnumSet` wherever possible and I am looking at ways to fix this.

**Comment by remkop** [comment]
Note to self: potential fix:

```
Index: src/main/java/picocli/CommandLine.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/main/java/picocli/CommandLine.java	(date 1550209901374)
+++ src/main/java/picocli/CommandLine.java	(date 1550209901374)
@@ -8803,7 +8803,12 @@
             List<Object> converted = consumeArguments(argSpec, lookBehind, arity, args, type, argDescription);
             if (collection == null || (!collection.isEmpty() && !initialized.contains(argSpec))) {
                 tracer.debug("Initializing binding for %s with empty %s%n", optionDescription("", argSpec, 0), argSpec.type().getSimpleName());
-                collection = createCollection(argSpec.type()); // collection type
+                if (collection instanceof EnumSet) {
+                    Object enumSet = EnumSet.noneOf((Class<Enum>) type);
+                    collection = (Collection<Object>) enumSet;
+                } else {
+                    collection = createCollection(argSpec.type()); // collection type
+                }
                 argSpec.setValue(collection);
             }
             initialized.add(argSpec);
@@ -9005,7 +9010,7 @@
                 return new ArrayList<Object>();
             }
             // custom Collection implementation class must have default constructor
-            return (Collection<Object>) collectionClass.newInstance();
+            return (Collection<Object>) factory.create(collectionClass);
         }
         @SuppressWarnings("unchecked") private Map<Object, Object> createMap(Class<?> mapClass) throws Exception {
             try { // if it is an implementation class, instantiate it
```

**Comment by leeatkinson** [comment]
Many thanks for your prompt response and help.

I've used your workaround and, because I need to set a default value, rather than assigning it in the code (which causes the exception) I can use a IDefaultValueProvider.

**Comment by remkop** [comment]
Fixed in master and the 3.9.x branch. Please verify.
I plan to release 3.9.4 including this fix soon.

**Comment by remkop** [comment]
Picocli 3.9.4 with a fix for this issue has been released. Enjoy!
